packages/block-editor/src/hooks/color.js

```diff
@@ -0,0 +1,250 @@
+		return addSaveProps( props, settings, attributes );
+	};
+
+	return settings;
+}
+
+const ColorPanel = ( { colorSettings, clientId } ) => {
+	const { getComputedStyle, Node } = window;
+
+	const [ detectedBackgroundColor, setDetectedBackgroundColor ] = useState();
+	const [ detectedColor, setDetectedColor ] = useState();
+
+	// TODO: check performance impact of running this on each render
+	useEffect( () => {
+		const colorsDetectionElement = getBlockDOMNode( clientId );
+		setDetectedColor( getComputedStyle( colorsDetectionElement ).color );
+
+		let backgroundColorNode = colorsDetectionElement;
+		let backgroundColor = getComputedStyle( backgroundColorNode )
+			.backgroundColor;
+		while (
+			backgroundColor === 'rgba(0, 0, 0, 0)' &&
+			backgroundColorNode.parentNode &&
+			backgroundColorNode.parentNode.nodeType === Node.ELEMENT_NODE
+		) {
+			backgroundColorNode = backgroundColorNode.parentNode;
+			backgroundColor = getComputedStyle( backgroundColorNode )
+				.backgroundColor;
+		}
+
+		setDetectedBackgroundColor( backgroundColor );
+	} );
+
+	return (
+		<InspectorControls>
+			<PanelColorSettings
+				title={ __( 'Color settings' ) }
+				initialOpen={ false }
+				colorSettings={ colorSettings }
+			>
+				<ContrastChecker
+					backgroundColor={ detectedBackgroundColor }
+					textColor={ detectedColor }
+				/>
+			</PanelColorSettings>
+		</InspectorControls>
+	);
+};
+
+/**
+ * Override the default edit UI to include new inspector controls for block
+ * color, if block defines support.
+ *
+ * @param  {Function} BlockEdit Original component
+ * @return {Function}           Wrapped component
+ */
+export const withBlockControls = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { name: blockName } = props;
+		const colors = useSelect( ( select ) => {
+			return select( 'core/block-editor' ).getSettings().colors;
+		}, [] );
+
+		if ( ! hasBlockSupport( blockName, COLOR_SUPPORT_KEY ) ) {
+			return <BlockEdit key="edit" { ...props } />;
+		}
+		const { style, textColor, backgroundColor } = props.attributes;
+
+		const onChangeColor = ( name ) => ( value ) => {
+			const colorObject = getColorObjectByColorValue( colors, value );
+			const attributeName = name + 'Color';
+			const newStyle = {
+				...style,
+				color: {
+					...style?.color,
+					[ name ]:
+						colorObject && colorObject.slug ? undefined : value,
+				},
+			};
+			const newNamedColor =
+				colorObject && colorObject.slug ? colorObject.slug : undefined;
+			props.setAttributes( {
+				style: newStyle,
+				[ attributeName ]: newNamedColor,
+			} );
+		};
+
+		return [
+			<ColorPanel
+				key="colors"
+				clientId={ props.clientId }
+				colorSettings={ [
+					{
+						label: __( 'Text Color' ),
+						onChange: onChangeColor( 'text' ),
+						colors,
+						value: getColorObjectByAttributeValues(
+							colors,
+							textColor,
+							style?.color?.text
+						).color,
+					},
+					{
+						label: __( 'Background Color' ),
```

Comment:
In cases that support gradient backgrounds, the UI for gradients should appear in the same panel as colors, should these hooks also handle gradients?
It would just require that we check for gradient support. Then we could unconditionally use PanelColorGradientSettings instead of PanelColorSettings and if gradients are supported we would pass different settings to the component.
I guess it could work well for the button to cover the logic is a little bit more complex as when gradients are added together with an image we need to insert a new dom element and the gradient there so I guess for cover we will always need some custom code.

Reply:
Yes, I wonder if we should auto-support gradients for all blocks that support background colors but regardless, I believe this hook should support gradients in the future.

Reply:
About whether we should auto-add the sub-element when gradients are used, ideally no but we might have to, the goal is to always be generic though as we need to be able to support consistent features across blocks way more quickly than what we do today. If we sacrifice some flexibility sometimes, it's fine.